connectors/migrations/20241212_backfill_github_data_source_folders.ts

```diff
@@ -0,0 +1,236 @@
+    const repoWithDiscussions = await getUniqueRepositoryIdsWithDiscussions(
+      connector.id
+    );
+    await concurrentExecutor(
+      repoWithDiscussions,
+      async (repoId) => {
+        // The node id is the same as the repoId
+        await upsertDiscussionFolderNode(
+          repoId,
+          getRepositoryInternalId(repoId),
+          dataSourceConfig
+        );
+      },
+      { concurrency: 16 }
+    );
+  }
+}
+
+async function upsertRepositoryFolderNode(
+  repository: GithubCodeRepository,
+  dataSourceConfig: DataSourceConfig
+) {
+  const repoFolderId = repository.repoId;
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: getRepositoryInternalId(repository.repoId),
+    parents: [repoFolderId],
+    title: repository.repoName,
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_FULL"),
+  });
+  return repoFolderId;
+}
+
+async function upsertCodeFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const codeFolderId = getCodeRootInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: codeFolderId,
+    parents: [codeFolderId, repositoryNodeId],
+    title: "Code",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_CODE"),
+  });
+  return codeFolderId;
+}
+
+async function upsertDirectoryFolderNode(
+  directory: GithubCodeDirectory,
+  repositoryId: number,
+  dataSourceConfig: DataSourceConfig
+) {
+  // This already contains IDs for Code and Repository folders
+  const parents = await getGithubCodeDirectoryParentIds(
+    directory.connectorId,
+    directory.internalId,
+    repositoryId
+  );
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: directory.internalId,
+    parents: [directory.internalId, ...parents],
+    title: directory.dirName,
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_CODE_DIR"),
+  });
+}
+
+async function upsertIssueFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const issuesFolderId = getIssuesInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: issuesFolderId,
+    parents: [issuesFolderId, repositoryNodeId],
+    title: "Issues",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_ISSUES"),
+  });
+}
+
+async function upsertDiscussionFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const discussionsFolderId = getDiscussionsInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: discussionsFolderId,
+    parents: [discussionsFolderId, repositoryNodeId],
+    title: "Discussions",
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_DISCUSSIONS"),
+  });
+}
+
+async function repositoryContainsCode(repository: GithubCodeRepository) {
+  const directory = await GithubCodeDirectory.findOne({
```

Comment:
you should get this from `codeSyncEnabled` in `github_connector_states`, there is an super edge case where code sync is enabled without having a single file.

Reply:
The reasoning is that there is no reason to have a Code folder if you don't actually have children to filters, but fair point

Reply:
This is super edge + this is a migration anyway so all good 👍

Reply:
The workflow will create the folder anyway since code sync is not webhook-based

Reply:
I changed it to be the way you suggested 👍